### PR TITLE
[Merged by Bors] - vm: introduce genesis id to vm

### LINKED ISSUE #3504: vm: include genesisid into a tx hash that is signed
part of: https://github.com/spacemeshos/pm/issues/135
depends on: https://github.com/spacemeshos/go-spacemesh/issues/3505

- initialize vm with GenesisID
- add GenesisID to Context type
- prefix message that is hashed with GenesisID https://github.com/spacemeshos/go-spacemesh/blob/develop/genvm/templates/multisig/multisig.go#L39 in Verify 
- for sdk module: add WithGenesisID option and prefix hashed message 
  https://github.com/spacemeshos/go-spacemesh/blob/develop/genvm/sdk/multisig/tx.go#L108
  https://github.com/spacemeshos/go-spacemesh/blob/develop/genvm/sdk/options.go#L10

### PR DESCRIPTION
**Pull request by dshulyak**
closes: https://github.com/spacemeshos/go-spacemesh/issues/3504

motivation behind this change is a cross-network replay protection. it works consistently across default templates, but in general the implementation is a template specific and other templates may go a different route.

for all templates that will be released in genesis (singlesig, multisig, vesting) replay protection works as follows:
- signer prefixes encoded transaction body with genesis id, hashes the result and signs it 
```
sig = sign(sha256(genesisid || body))
payload = body || sig
```
- verifier receives payload of the transaction. decodes sig. prefixes body with genesisid, hashes it and passes to ed25519 verify function

### CODE REVIEW

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

File: genvm/core/types.go [hunk 1]
```diff
@@ -100,6 +102,7 @@ type Host interface {
 	Handler() Handler
 	Template() Template
 	Layer() LayerID
+	GetGenesisID() Hash20
```
Comment (countvonzero) [inline genvm/core/types.go:105]: how about an interface `Hash([]byte)` instead and the template gets to decide how to hash?

Comment (dshulyak) [reply genvm/core/types.go:105]: i prefer to use to use simple hash function instead of some interface method on the host

Comment (countvonzero) [reply genvm/core/types.go:105]: my rationale is that in the future custom templates don't necessarily care about genesis id, as long as the template can parse its own tx correctly.

File: systest/cluster/cluster.go [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package cluster
 
 import (
+	"crypto/sha256"
```
Comment (countvonzero) [inline systest/cluster/cluster.go]: "github.com/minio/sha256-simd"?

**Review by countvonzero — APPROVED** [review]

File: genvm/core/types.go (hunk 1 above)
Comment (dshulyak) [reply genvm/core/types.go:105]: it is up to template how to implement replay protection, they are not necessarily will use hashing for that

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**